site/_posts/2018-12-12-reinventing-the-proxy.md

```diff
@@ -0,0 +1,124 @@
+
+```ruby
+class Proxy
+  def initialize
+    @pairing = {}
+
+    Thread.new do
+      loop do
+        reads, writes, _ = IO.select(@sockets, @sockets, [], 0.1)
+        # Iterate through the reads
+        read.each do |reads|
+          # Retrieve the other end and the translator method from the pairing hash
+          write, translate = @pairing[read]
+
+          # Skip the transmission if the other end is not writable
+          next unless writes.include?(writes)
+
+          # Do the transmission in a non-blocking way
+          data = read.read_nonblock(1024)
+          write.write_nonblock(translate.call(data))
+        end
+      end
+    end
+  end
+
+  def push(ws_sock, console_sock)
+    @sockets.push(ws_sock, console_sock)
+
+    # Pair the sockets and the translation methods
+    @pairing[ws_sock] = [console_sock, method(:translate_ws)]
+    @pairing[console_sock] = [ws_sock, method(:translate_console)]
+  end
+end
+```
+
+This implementation is really close to what we used to have in the last two years, of course with additional mutexes and error handling. As you can see, the `IO.select` was utilized to determine if a socket is ready for reading and/or writing. However, this implementation can cause a [spinlock](https://en.wikipedia.org/wiki/Spinlock) that is encoded into the usage of `IO.select` inside the `loop`. The last argument of the `IO.select` method is a timeout which states how many seconds should the thread wait if no socket is ready. This should cause our thread to sleep when there is nothing to transmit and so the computer can do other useful stuff in the meantime. The problem with this is that the `IO.select` will return from its sleep whenever a single socket is ready for transmission. In other words, `IO.select` cannot be forced to wait longer for a pair of sockets to become ready for reading and writing at least in one direction.
+
+Even though the spinlock is just a probabilistic scenario, problems might appear by just using a single socket pair and the probability increases with the number of sockets. Let's say we have two sockets and none of them are ready for reading or writing. In this case the thread will sleep, probably for more than a single iteration. After some waiting time, one of the sockets becomes ready for writing and the `IO.select` returns from the sleep earlier. There is no data transfer, as the other endpoint is not ready for reading so the loop continues. In the next iterations the `IO.select` will immediately return with the same single writable socket and skip the transmission until the other socket becomes readable. This will cause the thread to spinlock, i.e. to waste CPU time for actively waiting something to happen. It will not slow down the remote consoles, but everything else running on the appliance.
```

Comment:
"to waste CPU time for actively waiting something to happen" -> to waste CPU time on actively waiting for something to happen